Lib/test/test_asyncio/test_eager_task_factory.py

```diff
@@ -0,0 +1,241 @@
+        async def run():
+            await asyncio.sleep(0.1)
+            t = self.loop.create_task(coro())
+            self.assertFalse(t.done())
+            result = await t
+            self.assertTrue(t.done())
+            return result
+
+        result = self.loop.run_until_complete(run())
+        self.assertEqual(result, 'finished after blocking')
+
+    def test_cancellation_after_eager_completion(self):
+
+        async def coro():
+            return 'finished without blocking'
+
+        async def run():
+            await asyncio.sleep(0.1)
+            t = self.loop.create_task(coro())
+            t.cancel()
+            result = await t
+            # finished task can't be cancelled
+            self.assertFalse(t.cancelled())
+            return result
+
+        result = self.loop.run_until_complete(run())
+        self.assertEqual(result, 'finished without blocking')
+
+    def test_cancellation_after_eager_step_blocks(self):
+
+        async def coro():
+            await asyncio.sleep(0.1)
+            return 'finished after blocking'
+
+        async def run():
+            await asyncio.sleep(0.1)
+            t = self.loop.create_task(coro())
+            t.cancel('cancellation message')
+            self.assertGreater(t.cancelling(), 0)
+            try:
+                result = await t
+            except asyncio.CancelledError:
+                # finished task can't be cancelled
+                self.assertTrue(t.cancelled())
+                raise
+
+        with self.assertRaises(asyncio.CancelledError) as cm:
+            self.loop.run_until_complete(run())
+
+        self.assertEqual('cancellation message', cm.exception.args[0])
+
+
+class AsyncTaskCounter:
+    def __init__(self, loop, *, task_class, eager):
+        self.suspense_count = 0
+        self.task_count = 0
+
+        def CountingTask(*args, eager_start=False, **kwargs):
+            if not eager_start:
+                self.task_count += 1
+            kwargs["eager_start"] = eager_start
+            return task_class(*args, **kwargs)
+
+        if eager:
+            factory = asyncio.create_eager_task_factory(CountingTask)
+        else:
+            def factory(loop, coro, **kwargs):
+                return CountingTask(coro, loop=loop, **kwargs)
+        loop.set_task_factory(factory)
+
+    def get(self):
+        return self.task_count
+
+
+async def awaitable_chain(depth):
+    if depth == 0:
+        return 0
+    return 1 + await awaitable_chain(depth - 1)
+
+
+async def recursive_taskgroups(width, depth):
+    if depth == 0:
+        return
+
+    async with asyncio.TaskGroup() as tg:
+        futures = [
+            tg.create_task(recursive_taskgroups(width, depth - 1))
+            for _ in range(width)
+        ]
+
+
+async def recursive_gather(width, depth):
+    if depth == 0:
+        return
+
+    await asyncio.gather(
+        *[recursive_gather(width, depth - 1) for _ in range(width)]
+    )
+
+
+class BaseTaskCountingTests:
```

Comment:
Can this derive from `test_utils.TestCase`, allowing us to skip the multiple inheritance for `NonEagerTests` etc.?

Reply:
not really.

it's the same pattern as in other similar cases (like tasks tests), and the reasoning is that the base class has `test_*` methods but if it inherits from a `TestCase` then these methods will be executed as tests on the base class which fails (not surprisingly).